### PR TITLE
feat: optimized compression and expanded bit rates

### PR DESCRIPTION
**Pull request by ddeadguyy**
https://github.com/nfrechette/acl/issues/353
https://github.com/nfrechette/acl/issues/373

**ACL_SJSON_FIX**

Fix that supports testing of sjson_file_type::raw_track_list, instead of only sjson_file_type::raw_clip.  

I needed this to test my edge case animation within ACL.

**ACL_COMPRESSION_OPTIMIZED**

Prevent low-magnitude channels from becoming constant when worst-case shell distance and object space distance are exceeded. 
Apply error correction after constant and default tracks are processed.
Propagate shell distance and object space distance through ancestors before compression.

Massive memory and quality upgrade.
In our use case, there's a **28%** reduction in memory overall(**33%** in humans, **22%** in creatures).
Our worst edge case used to compress in 30 seconds, and failed regression with an error of nearly 2 meters.  Now regression passes well within precision settings, and it **compresses 10x faster**.
Compression is faster generally, as seen in _py make.py -regression_test_:
```
uniformly_sampled_database.config.sjson: 04.04s -> 03.38s
uniformly_sampled_database_4kb.config.sjson: 04.03s -> 03.33s
uniformly_sampled_database_4kb_mixed.config.sjson: 04.73s  -> 04.04s
uniformly_sampled_database_mixed.config.sjson: 04.64s -> 03.99s
uniformly_sampled_mixed_var_0.config.sjson: 01.14s -> 01.26s
uniformly_sampled_mixed_var_1.config.sjson: 01.61s -> 01.35s
uniformly_sampled_quant_bind_relative.config.sjson: 01.87s -> 01.29s
uniformly_sampled_quant_high.config.sjson: 02.27s -> 01.33s
uniformly_sampled_quant_highest.config.sjson: 03.82s -> 01.27s
uniformly_sampled_quant_medium.config.sjson: 01.68s -> 01.27s
uniformly_sampled_quant_mtx_error.config.sjson: 01.68s -> 01.27s
uniformly_sampled_raw.config.sjson: 01.19s -> 01.20s
```

I also tried a more accurate version of propagation which compared every bone with every ancestor.  It resulted in smaller distances, but was more complex(O(N^2)), compression took longer, and memory savings shrank.

I'm confident that introducing error correction within segment compression would improve memory and compression time further.  It seems unlikely that most of the logic in find_optimal_bit_rates would be required anymore.  Perhaps I'll experiment with this later, but the current state of ACL_COMPRESSION_OPTIMIZED is fast enough, and high-quality enough, for our immediate needs.

**ACL_BIT_RATE_EXPANSION**

Expand bit rate options, from [3..19] to [1..23].  

Note that these results assume that ACL_COMPRESSION_OPTIMIZED is enabled.  I don't recommend trying ACL_BIT_RATE_EXPANSION without it.
Additional 2% reduction in memory overall(3% in humans, 2% in creatures).
No changes to _py make.py -regression_test_ timings worth reporting, but some of our creature compression does get slower.  Edge cases are rare, and max out at 7 seconds this time.

### CODE REVIEW

**Comment by ddeadguyy** [comment]
It's worth noting that ACL_BIT_RATE_EXPANSION does slow down unit tests, but ACL_UNIT_TEST_STRICT from https://github.com/nfrechette/acl/pull/369 helps with that.

File: includes/acl/compression/impl/clip_context.h [hunk 1]
```diff
@@ -137,6 +137,17 @@ namespace acl
...
 			uint32_t* leaf_transform_chains				= nullptr;
 
+#ifdef ACL_COMPRESSION_OPTIMIZED
+
+			struct transform_link
+			{
+				uint32_t parent_transform_index;
+				uint32_t child_transform_index;
+			};
+			transform_link* transform_links				= nullptr;	// 1 per transform, sorted in reverse order, for compact_constant_streams and quantization_context.adjusted_shell_distances.
```
Comment (nfrechette) [inline includes/acl/compression/impl/clip_context.h:147]: Reserve order = children first?

File: includes/acl/compression/impl/clip_context.h [hunk 2]
```diff
@@ -309,6 +337,34 @@ namespace acl
 				uint32_t num_root_bones = 0;
 #endif
 
+#ifdef ACL_COMPRESSION_OPTIMIZED
+
+				if (transform_links != nullptr)
+				{
+					auto sort_predicate = [](const clip_context::transform_link& lhs, const clip_context::transform_link& rhs)
```
Comment (nfrechette) [inline includes/acl/compression/impl/clip_context.h:344]: Could benefit from a comment to clarify the sort order

File: includes/acl/compression/impl/compact_constant_streams.h [hunk 3]
```diff
@@ -77,12 +94,65 @@ namespace acl
...
+					continue;
+				}
+
+				const rtm::vector4f vtx0 = rtm::quat_mul_vector3(shell_point_x, rotation);
+				const rtm::vector4f vtx1 = rtm::quat_mul_vector3(shell_point_y, rotation);
+
+				const rtm::scalarf vtx0_error = rtm::vector_distance3(ref_vtx0, vtx0);
+				const rtm::scalarf vtx1_error = rtm::vector_distance3(ref_vtx1, vtx1);
+
+				if (rtm::scalar_cast(rtm::scalar_max(vtx0_error, vtx1_error)) > precision)
```
Comment (nfrechette) [inline includes/acl/compression/impl/compact_constant_streams.h:111]: Really good idea!

File: includes/acl/compression/impl/compact_constant_streams.h [hunk 4]
```diff
@@ -110,7 +253,22 @@ namespace acl
 				const float constant_translation_threshold = settings.translation_format != vector_format8::vector3f_full ? desc.constant_translation_threshold : 0.0F;
 				const float constant_scale_threshold = settings.scale_format != vector_format8::vector3f_full ? desc.constant_scale_threshold : 0.0F;
 
-				if (is_rotation_track_constant(bone_stream.rotations, constant_rotation_threshold_angle))
+				if (
+
+#ifdef ACL_COMPRESSION_OPTIMIZED
+
+					bone_stream.is_rotation_constant || bone_range.rotation.is_constant() || ((constant_rotation_threshold_angle > 0.0F) &&
```
Comment (nfrechette) [inline includes/acl/compression/impl/compact_constant_streams.h:260]: Is the check for bone_stream_is_rotation_constant necessary? It should only be true here if we have a single sample (from the context init).

File: includes/acl/compression/impl/compact_constant_streams.h [hunk 5]
```diff
@@ -149,10 +342,195 @@ namespace acl
...
+#endif
+
 				}
 			}
 
 			context.has_scale = num_default_bone_scales != num_bones;
+
+#ifdef ACL_COMPRESSION_OPTIMIZED
+
+			deallocate_type_array(allocator, max_adjusted_shell_distances, num_bones);
```
Comment (nfrechette) [inline includes/acl/compression/impl/compact_constant_streams.h:360]: Could benefit from a few comments to explain what is going on.
If I'm not mistaken, we apply error correction to our constant sub-tracks by transforming everything into object space and back into local space.

File: includes/acl/compression/impl/compact_constant_streams.h [hunk 6]
```diff
@@ -149,10 +342,195 @@ namespace acl
...
+							{
+								constant_changed.scale = parent_constant_changed.scale;
+							}
+
+							// Compensate for the constant changes in your ancestors.
+							if (!bone_stream.is_rotation_constant && parent_constant_changed.rotation)
+							{
+								ACL_ASSERT(any_constant_changed.rotation, "No rotations have changed!");
+								adjusted_local_transform.rotation = rtm::quat_normalize(rtm::quat_mul(original_object_transform.rotation, rtm::quat_conjugate(parent_adjusted_object_transform.rotation)));
+								raw_bone_stream.rotations.set_raw_sample(sample_index, adjusted_local_transform.rotation);
```
Comment (nfrechette) [inline includes/acl/compression/impl/compact_constant_streams.h:478]: I'm not sure if we should modify the raw data. By doing so, any error we measure afterwards will no longer reflect what the true error is compared to the original raw clip. For example, a leaf bone might end up moved through error correction and even if it is constant, in local space, it would report an error of 0cm even though it no longer matches the original raw value.

But, if we don't modify the raw value accordingly, the bit rate optimization might try to reach a target that is unreachable... this might lead to more bits being used in vain.

Either way, we need to add a comment here to clarify why we are doing this as it is not obvious.

File: includes/acl/compression/impl/track_error.impl.h [hunk 7]
```diff
@@ -641,6 +641,8 @@ namespace acl
 		return calculate_transform_track_error(allocator, args);
 	}
 
+#ifndef ACL_COMPRESSION_OPTIMIZED
```
Comment (nfrechette) [inline includes/acl/compression/impl/track_error.impl.h:644]: Why strip this?

File: includes/acl/compression/impl/write_stream_data.h [hunk 8]
```diff
@@ -331,8 +331,19 @@ namespace acl
 				}
 				else
 				{
+
+#ifdef ACL_BIT_RATE_EXPANSION
+
+					const uint64_t* raw_sample_u64 = safe_ptr_cast<const uint64_t>(raw_sample_ptr);
```
Comment (nfrechette) [inline includes/acl/compression/impl/write_stream_data.h:337]: Aren't these equivalent?

**Review by nfrechette — COMMENTED** [review]
Added a few notes and questions

**Comment by nfrechette** [comment]
Wow, very impressive savings!
I wonder if this mostly happens because we now modify the raw data to match the constant value which wasn't the case before. This would have left the bit rate optimization trying to reach a target that is unreachable even with full float precision. This would have lead to more higher bit rates being used and more bit rates tested which would explain the memory savings and the compression time speedup. The error seen by the quantization being lower now makes sense since the algorithm has room to actually reach the tweaked raw value. However, it wouldn't lead to a true lower error. The error correction probably helps here.

File: includes/acl/compression/impl/clip_context.h (hunk 1 above)
Comment (ddeadguyy) [reply includes/acl/compression/impl/clip_context.h:147]: Fair point, I can be clearer about what "reverse order" means.

File: includes/acl/compression/impl/compact_constant_streams.h (hunk 4 above)
Comment (ddeadguyy) [reply includes/acl/compression/impl/compact_constant_streams.h:260]: I actually considered removing all of the constant and default processing from constant init, so we can do it all here and exert maximum control over tolerance tests(such as avoiding <= and < differences).  I left it alone just to keep the number of edits down.  Having said that, I can still remove the bone_stream.is_X_constant checks, because they're covered by the new ultra-precise bone_range.X.is_constant() calls anyhow.

File: includes/acl/compression/impl/compact_constant_streams.h (hunk 5 above)
Comment (ddeadguyy) [reply includes/acl/compression/impl/compact_constant_streams.h:360]: hat's the gist of it. We use object space of the original data as ground truth, and only deviate for 2 reasons, and as briefly as possible.

Replace an original local value with a new constant value.
Correct for the manipulation of an original local value by an ancestor ASAP.

File: includes/acl/compression/impl/compact_constant_streams.h (hunk 6 above)
Comment (ddeadguyy) [reply includes/acl/compression/impl/compact_constant_streams.h:478]: Another fair point on the comment.

We aren't modifying raw data here.  We're modifying the raw channels generated from the raw data.  The raw data is left alone, and is still used at the end of the process to do regression testing.  Regression results are much tighter with this change than without.  My worst-case animation doesn't fail with a 2m error anymore(and doesn't glitch ingame anymore either).  It settles nicely near the 0.001 mark, which matches compression precision.  I suspect that regression tolerance can be reduced to compression precision now, as opposed to what it's been using.


>  bit rate optimization might try to reach a target that is unreachable... this might lead to more bits being used in vain.

Correct.  This is what was happening before ACL_COMPRESSION_OPTIMIZED.

File: includes/acl/compression/impl/track_error.impl.h (hunk 7 above)
Comment (ddeadguyy) [reply includes/acl/compression/impl/track_error.impl.h:644]: They were unused.  They also didn't seem to make much sense, such as _"Calculates the worst compression error between two raw track arrays."_  I wouldn't expect raw track arrays to have any compression at all, never mind compression error.  It was part of my initial review of the compression system, but I can put them back if you prefer.

File: includes/acl/compression/impl/write_stream_data.h (hunk 8 above)
Comment (ddeadguyy) [reply includes/acl/compression/impl/write_stream_data.h:337]: Not quite.  This is the change I tracked down to fix high bitrate regression failures in 1.3.
`const uint64_t raw_sample_u64 = *safe_ptr_cast<const uint64_t>(raw_sample_ptr);`
This copies 64 bits of memory into a local 64 bit value.  Then the address of that local value is passed to the next stage.  This only works if num_bits_at_bit_rate <= 64, otherwise the last few bits are junk.

**Comment by nfrechette** [comment]
It looks like the debug unit tests have exploded in execution time.
Some runs now take 40mins+ where before everything ran in <10mins.
For example, this run from a few days ago is normal: https://ci.appveyor.com/project/nfrechette/acl/builds/39556495

These are the ones that seem much worse:
31/43 Test #28: unpack_scalarf_uXX_unsafe .................   Passed  496.69 sec
38/43 Test #32: pack_vector4_XX ...........................   Passed  542.31 sec
41/43 Test #42: pack_vector2_XX ...........................   Passed  499.81 sec
43/43 Test #39: pack_vector3_XX ...........................   Passed  815.68 sec

**Comment by ddeadguyy** [comment]
Yeah, ACL_BIT_RATE_EXPANSION cranks them way up.  It isn't so bad in release(at least not on my machine).  That's what I was referring to in this comment.

> It's worth noting that ACL_BIT_RATE_EXPANSION does slow down unit tests, but ACL_UNIT_TEST_STRICT from #369 helps with that.

ACL_UNIT_TEST_STRICT leverages SIMD to make vector2 ~2x faster, vector3 ~3x faster, and vector4 ~4x faster than scalar.  That could be enough to get debug unit tests under the limit.

**Comment by ddeadguyy** [comment]
> I could manually bring in the SIMD-leveraging parts though, without making the tests stricter, with comments to replace them with ACL_UNIT_TEST_STRICT in 2.1.

Looks like this plan worked.

**Comment by nfrechette** [comment]
Awesome stuff, thanks!
